roachtest: sqlsmith/setup=empty/setting=default failed

[(roachtest).sqlsmith/setup=empty/setting=default failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2071705&tab=buildLog) on [provisional_202007071743_v20.2.0-alpha.2@0b6e118bc1bcba4cfb4fc6c660153ec5be3989e8](https://github.com/cockroachdb/cockroach/commits/0b6e118bc1bcba4cfb4fc6c660153ec5be3989e8):

```
The test failed on branch=provisional_202007071743_v20.2.0-alpha.2, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=empty/setting=default/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:757: error: pq: internal error: inconsistent Case return types bool unknown
		stmt:
		WITH
			with_60 (col_342)
				AS (
					SELECT
						'-2000-01-01 00:00:00+00:00':::TIMESTAMP AS col_342
					FROM
						(
							VALUES
								(true, (SELECT 2409586731:::OID AS col_339 LIMIT 1:::INT8)),
								(true, 276957771:::OID),
								(
									CASE
									WHEN st_overlaps(st_geomfromgeojson('pw':::STRING::STRING)::GEOMETRY::GEOMETRY, '0101000000000000000000F03F000000000000F03F':::GEOMETRY::GEOMETRY)::BOOL
									THEN false
									ELSE (NULL AND true)
									END,
									3378038347:::OID
								),
								(NULL, 3587599014:::OID),
								(NULL, 3142601850:::OID)
						)
							AS tab_203 (col_340, col_341)
				)
		SELECT
			tab_204.col_343 AS col_344
		FROM
			(VALUES ('':::STRING), ('':::STRING), (NULL), (NULL), ('U>%"~N':::STRING)) AS tab_204 (col_343)
		WHERE
			true
		ORDER BY
			tab_204.col_343 ASC
		LIMIT
			1:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=default](https://teamcity.cockroachdb.com/viewLog.html?buildId=2071705&tab=artifacts#/sqlsmith/setup=empty/setting=default)
Related:
- #47541 roachtest: sqlsmith/setup=empty/setting=default failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Ddefault.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/typing
+++ b/pkg/sql/logictest/testdata/logic_test/typing
@@ -243,3 +243,16 @@ query T
 SELECT CASE WHEN true THEN 1234:::OID ELSE COALESCE(NULL, NULL) END
 ----
 1234
+
+# Regression test for #51099.
+query B rowsort
+SELECT CASE WHEN x > 1 THEN true ELSE NULL AND true END FROM (VALUES (1), (2)) AS v(x)
+----
+NULL
+true
+
+query B rowsort
+SELECT CASE WHEN x > 1 THEN true ELSE NULL OR false END FROM (VALUES (1), (2)) AS v(x)
+----
+NULL
+true

--- a/pkg/sql/opt/norm/testdata/rules/bool
+++ b/pkg/sql/opt/norm/testdata/rules/bool
@@ -285,8 +285,8 @@ project
  │    ├── columns: k:1!null
  │    └── key: (1)
  └── projections
-      ├── NULL OR (k:1 = 1) [as=r:6, outer=(1)]
-      └── NULL AND (k:1 = 1) [as=s:7, outer=(1)]
+      ├── CAST(NULL AS BOOL) OR (k:1 = 1) [as=r:6, outer=(1)]
+      └── CAST(NULL AS BOOL) AND (k:1 = 1) [as=s:7, outer=(1)]
 
 # --------------------------------------------------
 # FoldNotTrue + FoldNotFalse + FoldNotNull
@@ -667,7 +667,7 @@ project
  ├── fd: ()-->(6)
  ├── scan a
  └── projections
-      └── NULL [as=r:6]
+      └── CAST(NULL AS BOOL) [as=r:6]
 
 norm expect=(ExtractRedundantConjunct)
 SELECT (null and k=2) or (null and k=1) AS r FROM a
@@ -678,7 +678,7 @@ project
  │    ├── columns: k:1!null
  │    └── key: (1)
  └── projections
-      └── NULL AND ((k:1 = 2) OR (k:1 = 1)) [as=r:6, outer=(1)]
+      └── CAST(NULL AS BOOL) AND ((k:1 = 2) OR (k:1 = 1)) [as=r:6, outer=(1)]
 
 # Check that we don't match non-redundant cases.
 norm expect-not=(ExtractRedundantConjunct)

--- a/pkg/sql/opt/optbuilder/scalar.go
+++ b/pkg/sql/opt/optbuilder/scalar.go
@@ -113,8 +113,8 @@ func (b *Builder) buildScalar(
 		return b.finishBuildScalarRef(t.col, inScope, outScope, outCol, colRefs)
 
 	case *tree.AndExpr:
-		left := b.buildScalar(t.TypedLeft(), inScope, nil, nil, colRefs)
-		right := b.buildScalar(t.TypedRight(), inScope, nil, nil, colRefs)
+		left := b.buildScalar(tree.ReType(t.TypedLeft(), types.Bool), inScope, nil, nil, colRefs)
+		right := b.buildScalar(tree.ReType(t.TypedRight(), types.Bool), inScope, nil, nil, colRefs)
 		out = b.factory.ConstructAnd(left, right)
 
 	case *tree.Array:
@@ -311,7 +311,7 @@ func (b *Builder) buildScalar(
 		out = b.factory.ConstructVariable(inScope.cols[t.Idx].id)
 
 	case *tree.NotExpr:
-		input := b.buildScalar(t.TypedInnerExpr(), inScope, nil, nil, colRefs)
+		input := b.buildScalar(tree.ReType(t.TypedInnerExpr(), types.Bool), inScope, nil, nil, colRefs)
 		out = b.factory.ConstructNot(input)
 
 	case *tree.IsNullExpr:
@@ -345,8 +345,8 @@ func (b *Builder) buildScalar(
 		out = b.factory.ConstructCase(input, whens, input)
 
 	case *tree.OrExpr:
-		left := b.buildScalar(t.TypedLeft(), inScope, nil, nil, colRefs)
-		right := b.buildScalar(t.TypedRight(), inScope, nil, nil, colRefs)
+		left := b.buildScalar(tree.ReType(t.TypedLeft(), types.Bool), inScope, nil, nil, colRefs)
+		right := b.buildScalar(tree.ReType(t.TypedRight(), types.Bool), inScope, nil, nil, colRefs)
 		out = b.factory.ConstructOr(left, right)
 
 	case *tree.ParenExpr:

--- a/pkg/sql/opt/optbuilder/testdata/scalar
+++ b/pkg/sql/opt/optbuilder/testdata/scalar
@@ -1259,3 +1259,27 @@ case [type=oid]
       │    └── null [type=unknown]
       └── cast: OID [type=oid]
            └── null [type=unknown]
+
+# Make sure NULL arguments for AND/OR/NOT are typed as boolean.
+build-scalar
+NULL AND true
+----
+and [type=bool]
+ ├── cast: BOOL [type=bool]
+ │    └── null [type=unknown]
+ └── true [type=bool]
+
+build-scalar
+false OR NULL
+----
+or [type=bool]
+ ├── false [type=bool]
+ └── cast: BOOL [type=bool]
+      └── null [type=unknown]
+
+build-scalar
+NOT NULL
+----
+not [type=bool]
+ └── cast: BOOL [type=bool]
+      └── null [type=unknown]

